app/models/rental.rb

```diff
@@ -92,34 +110,77 @@ def renter_is_assignable
     end
   end
 
-  def create_reservation
-    throw :abort unless valid? # check if the current rental object is valid or not
+  # Note, rspec will mock this so we wont communicate with the api
+  def create_reservations
+    # if we fail half way through creating reservations we should roll them all back
+    created_reservations = []
     begin
-      reservation = Inventory.create_reservation(item_type.name, start_time, end_time)
-      raise 'Reservation UUID was not present in response.' unless reservation[:uuid].present?
+      raise RecordInvalid, 'Rental invalid' unless valid?
+      rentals_items.each do |ri|
+        raise RecordInvalid, 'Rental item is invalid' unless ri.valid? # check if the current rental item is valid
+
+        reservation = Inventory.create_reservation(ri.item_type.name, start_time, end_time)
+        raise 'Reservation UUID was not present in response.' unless reservation[:uuid].present?
 
-      self.reservation_id = reservation[:uuid]
-      self.item = Item.find_by(name: reservation[:item][:name])
+        ri.reservation_id = reservation[:uuid]
+        created_reservations << reservation[:uuid]
+        ri.item = Item.find_by(name: reservation[:item][:name])
+      end
     rescue => error
-      errors.add :base, error.inspect
+      # roll back the rentals we got partially through creating
+      failed_roll_back = false
+      created_reservations.each do |uuid|
+        # remove reservation from RentalsItem
+        if (ri = rentals_items.to_a.find { |_ri| _ri.reservation_id == uuid })
```

Comment:
It's better not to do variable assignment in a condition, but to do it before and then just write your condition using that variable.